daemon/monitor.go

```diff
@@ -85,16 +85,29 @@ func (daemon *Daemon) StateChanged(id string, e libcontainerd.StateInfo) error {
 		}
 		return daemon.postRunProcessing(c, e)
 	case libcontainerd.StateExitProcess:
+		closeAsync := make(chan bool)
+		waitClose := false
 		c.Lock()
-		defer c.Unlock()
-		if execConfig := c.ExecCommands.Get(e.ProcessID); execConfig != nil {
-			ec := int(e.ExitCode)
-			execConfig.ExitCode = &ec
-			execConfig.Running = false
-			execConfig.Wait()
-			if err := execConfig.CloseStreams(); err != nil {
-				logrus.Errorf("%s: %s", c.ID, err)
+		defer func() {
+			c.Unlock()
+			if waitClose {
+				<-closeAsync
 			}
+		}()
+		if execConfig := c.ExecCommands.Get(e.ProcessID); execConfig != nil {
+			waitClose = true
+			go func() {
+				ec := int(e.ExitCode)
+				execConfig.Lock()
+				defer execConfig.Unlock()
+				execConfig.ExitCode = &ec
+				execConfig.Running = false
+				execConfig.Wait()
+				if err := execConfig.CloseStreams(); err != nil {
+					logrus.Errorf("%s: %s", c.ID, err)
+				}
+				closeAsync <- true
+			}()
```

Comment:
The reason that `c.ExecCommands.Delete(execConfig.ID)` isn't in here is that it looked like the container lock should be held when deleting it (which the go routine is added to stop holding).

On a second look though, the ExecCommands store has its own lock. Do we even need to grab the container lock here at all? It doesn't look like anywhere else that mofidies the ExecCommands grabs the container lock. If not, this could be a much cleaner fix.